doc/decisions/1_in_discussion/internal_cache.md

```diff
@@ -0,0 +1,587 @@
+      if `dest->data->size` == 0 (no keys are in this keyset):
+        deallocate `dest->data` and point `dest->data` to `source->data`, increase `source->data->refs` by 1.
+      else if `dest->data->size` > 0 (meaning there are already keys in this keyset):
+        copy the keys of `source->data->array` over to `dest->data->array`
+    else if `dest->data->refs` > 0: (there are other keysets sharing the same data):
+      if `dest->data->size` == 0 (no keys are in this keyset):
+        decrease `dest->data->refs` by 1.
+        point `dest->data` to `source->data`, increase `source->data->refs` by 1.
+      else if `dest->data->size` > 0 (meaning there are already keys in this keyset):
+        decrease `dest->data->refs` by 1.
+        allocate new `dest->data`
+        copy the keys from old to `dest->data`
+        copy the keys of `source->data->array` over to `dest->data->array`
+  ```
+- `ksAppendKey(dest, key)`:
+
+  - Pretty much the same `ksAppend()` but only for a single key
+  - if `dest->data` is NULL, allocate `dest->data` and put the key there
+
+- `ksClear(dest)`:
+
+  ```
+  if `dest->data->refs` == 0 (meaning no other keysets points to it):
+    deletes the keys in `dest->data`
+  else if `dest->data->refs` > 0: (there are other keysets sharing the same data):
+    decrease `dest->data->refs` by 1
+    point `dest->data` to NULL
+  ```
+
+- `ksCut(dest, key)`:
+  ```
+  if `dest->data->refs` == 0 (meaning no other keysets points to it):
+    perform the cutting algorithm directly on these keys
+  else if `dest->data->refs` > 0: (there are other keysets sharing the same data):
+    decrease `dest->data->refs` by 1
+    create a new `dest->data`
+    copy over the non-cut keys into `dest->data` from the old `dest->data`
+  ```
+- `ksPop(dest)`:
+
+  ```
+  if `dest->data->refs` == 0 (meaning no other keysets points to it):
+    pop the last key directly on the keys
+  else if `dest->data->refs` > 0: (there are other keysets sharing the same data):
+    decrease `dest->data->refs` by 1
+    create a new `dest->data`
+    copy over all but the last keys from the old `dest->data`
+  ```
+
+- `ksLookup()`:
+  - if the `DEL` or `POP` flag is specified, do the COW-stuff as described multiple times now in the operations above
+
+#### Optimizations
+
+This approach requires more allocation than previously.
+We have not benchmarked whether this is big of an issue.
+One optimization could be an expanding "pool" of `_KeySetData`, `_KeyData` and `_KeyName`.
+We could then allocate multiple of them at the same time, and borrow and give back instance from and to the pool.
+
+## Memory comparison of COW approaches
+
+The following calculations are based on the AMD64 platform.
+All results are in bytes unless stated otherwise.
+
+Example key: `user:/hosts/ipv6/example.com = 2606:2800:220:1:248:1893:25c8:1946`
+
+We want to measure the following properties:
+
+- Empty KeySet: size of a simple malloc of the keyset struct
```

Comment:
Either remove the KeySet column, or add a second table similar to the one for Keys that fully shows the situation for KeySets, i.e., "Empty KeySet", "Empty KeySet + extra structs", "KeySet with 1 key", etc.

The current table suggests the "full-blown COW" needs less memory for a keyset, when this is not the case. It's only the case, when the keyset shares it's keys with another.

Reply:
Agree, I added another table for the keyset.